File: src/main/java/no/novari/flyt/file/FileService.java
package no.novari.flyt.file;

import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.fintlabs.cache.FintCache;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;

@Service
@AllArgsConstructor
@Slf4j
public class FileService {

    private final FintCache<UUID, File> fileCache;
    private final FileRepository fileRepository;

    public Mono<Optional<File>> findById(UUID fileId) {
        try {
            Optional<File> optionalFileFromCache = fileCache.getOptional(fileId);
            if (optionalFileFromCache.isPresent()) {
                return Mono.just(optionalFileFromCache);
            }
            return fileRepository.findById(fileId);
        } catch (Exception e) {
            return Mono.error(e);
        }
    }

    public Mono<UUID> put(UUID fileId, File file) {
        try {
            if (Objects.isNull(fileId)) {
                return Mono.error(new IllegalArgumentException("FileId cannot be null"));
            }
            if (Objects.isNull(file)) {
                return Mono.error(new IllegalArgumentException("File cannot be null"));
            }
            fileCache.put(fileId, file);
        } catch (Exception e) {
            return Mono.error(e);
        }
        try {
            return fileRepository.putFile(fileId, file)
                    .doOnError(e -> fileCache.remove(fileId))
                    .thenReturn(fileId);
        } catch (Exception e) {
            fileCache.remove(fileId);
            return Mono.error(e);
        }
    }

    public Mono<Void> delete(List<UUID> fileIds) {
        try {
            if (Objects.isNull(fileIds) || fileIds.isEmpty()) {
                log.info("List of fileIds is empty");
                return Mono.empty();
            }
            fileCache.remove(fileIds);
            return fileRepository.deleteFiles(fileIds);
        } catch (Exception e) {
            return Mono.error(e);
        }
    }

    public int deleteFilesOlderThan(int days) {
        return fileRepository.deleteFilesOlderThan(days);
    }

}

File: src/main/kotlin/no/novari/flyt/files/application/FileService.kt
package no.novari.flyt.files.application

import io.github.oshai.kotlinlogging.KotlinLogging
import no.novari.flyt.files.domain.FileDownload
import no.novari.flyt.files.domain.FileMetadata
import no.novari.flyt.files.domain.FilePayload
import no.novari.flyt.files.domain.exception.FileNotFoundException
import no.novari.flyt.files.infrastructure.storage.FileRepository
import org.springframework.stereotype.Service
import java.text.Normalizer
import java.util.UUID

@Service
class FileService(
    private val fileRepository: FileRepository,
) {
    private val log = KotlinLogging.logger {}

    fun openDownload(fileId: UUID): FileDownload {
        val fileDownload = fileRepository.openDownload(fileId) ?: throw FileNotFoundException(fileId)
        val normalizedMetadata = normalizeFileName(fileDownload.metadata)

        log.atDebug {
            message = "Resolved file metadata for fileId={} normalizedNameChanged={}"
            arguments =
                arrayOf(
                    fileId,
                    fileDownload.metadata.name != normalizedMetadata.name,
                )
        }

        return FileDownload(normalizedMetadata, fileDownload.contents)
    }

    fun put(
        fileId: UUID,
        file: FilePayload,
    ): UUID {
        val normalizedFile = normalizeFileName(file)

        log.atDebug {
            message = "Preparing file upload for fileId={} normalizedNameChanged={}"
            arguments =
                arrayOf(
                    fileId,
                    file.name != normalizedFile.name,
                )
        }

        return fileRepository.putFile(fileId, normalizedFile)
    }

    fun delete(fileIds: List<UUID>) {
        if (fileIds.isEmpty()) {
            log.atInfo {
                message = "List of fileIds is empty"
            }
            return
        }

        fileRepository.deleteFiles(fileIds)
    }

    fun deleteFilesOlderThan(days: Int): Int {
        return fileRepository.deleteFilesOlderThan(days)
    }

    private fun normalizeFileName(file: FileMetadata): FileMetadata {
        val normalizedName = normalizeFileName(file.name)

        return if (normalizedName == file.name) {
            file
        } else {
            file.copy(name = normalizedName)
        }
    }

    private fun normalizeFileName(file: FilePayload): FilePayload {
        val normalizedName = normalizeFileName(file.name)

        return if (normalizedName == file.name) {
            file
        } else {
            file.copy(name = normalizedName)
        }
    }

    private fun normalizeFileName(fileName: String): String {
        return fileName
            .trim()
            .replace(WHITESPACE_BEFORE_EXTENSION_REGEX, "$1")
            .replace(TYPOGRAPHIC_DASHES_REGEX, "-")
            .let { Normalizer.normalize(it, Normalizer.Form.NFC) }
    }

    companion object {
        private val WHITESPACE_BEFORE_EXTENSION_REGEX = Regex("""\s+(\.[^.\s]+(?:\.[^.\s]+)*)$""")
        private val TYPOGRAPHIC_DASHES_REGEX = Regex("[\u2013\u2014]")
    }
}
